ddl/check.lua

```diff
@@ -628,75 +628,151 @@ local function check_index(i, index, space)
+        )
+    end
 
-        if space.indexes.bucket_id ~= nil then
-            return nil, string.format(
-                "spaces[%q].indexes[%q]: bucket_id is used for sharding, " ..
-                "but there's no spaces[%q].sharding_key defined",
-                space.name, 'bucket_id', space.name
-            )
-        end
+    if space.fields.bucket_id.type ~= 'unsigned' then
+        return nil, string.format(
+            "spaces[%q].format[%q].type: bad value (unsigned expected, got %s)",
+            space.name, 'bucket_id', space.fields.bucket_id.type
+        )
+    end
+
+    return true
+end
+
+local function check_bucket_id_index(space)
+    if not space.indexes.bucket_id then
+        return nil, string.format(
+            "spaces[%q].indexes: sharding_key exists in the space, but there's" ..
+            " no bucket_id defined in 'indexes' section",
+            space.name
+        )
+    end
+
+    if space.indexes.bucket_id.unique then
+        return nil, string.format(
+            "spaces[%q].indexes[%q].unique: bucket_id index can't be unique",
+            space.name, 'bucket_id'
+        )
+    end
+
+    if #space.indexes['bucket_id'].parts ~= 1 then
+        return nil, string.format(
+            "spaces[%q].indexes[%q].parts: bucket_id index can't be composite (1 part expected, got %d parts)",
+            space.name, 'bucket_id', #space.indexes['bucket_id'].parts
+        )
+    end
+
+    if space.indexes['bucket_id'].parts[1].path ~= 'bucket_id' then
+        return nil, string.format(
+            "spaces[%q].indexes[%q].parts[1].path: invalid field reference " ..
+            "(reference to bucket_id expected, got %s)",
+            space.name, 'bucket_id', space.indexes['bucket_id'].parts[1].path
+        )
+    end
+
+    return true
+end
+
+local function is_callable(object)
+    if type(object) == 'function' then
         return true
     end
 
-    do -- check that format['bucket_id'] valid
-        if not space.fields.bucket_id then
-            return nil, string.format(
-                "spaces[%q].format: sharding_key exists in the space, but there's" ..
-                " no bucket_id defined in 'format' section",
-                space.name
-            )
-        end
+    -- all objects with type `cdata` are allowed
+    -- because there is no easy way to get
+    -- metatable.__call of object with type `cdata`
+    if type(object) == 'cdata' then
+        return true
+    end
 
-        if space.fields.bucket_id.type ~= 'unsigned' then
-            return nil, string.format(
-                "spaces[%q].format[%q].type: bad value (unsigned expected, got %s)",
-                space.name, 'bucket_id', space.fields.bucket_id.type
-            )
-        end
+    -- all objects with type `userdata` are allowed
+    -- because there is no easy way to get
+    -- metatable.__call of object with type `userdata`
+    if type(object) == 'userdata' then
+        return true
+    end
```

Comment:
`getmetatable()` works for a userdata in the same way as for a table. However there is a weird case about both userdata and table objects: if `__metatable` field is defined within a metatable, we unable to get actual metatable to check its `__call` field.

```lua
tarantool> x = setmetatable({}, {__call = function(self, ...) return ... end, __metatable = 'hidden'})
tarantool> getmetatable(x)
---
- hidden
...
tarantool> x(1, 2, 3)
---
- 1
- 2
- 3
...
```

However there is a trick: we can use `debug.getmetatable()`.

```lua
tarantool> debug.getmetatable(x)
---
- __metatable: hidden
  __call: 'function: 0x403d1130'
...
```

It is not nice to rely on the 'debug' library in production code. I'll cite Mike Pall (from [here](https://github.com/LuaJIT/LuaJIT/issues/536#issuecomment-565797780)):

> The purpose of the `debug.*` functions is for debugging Lua code. Period. In general, you'd want to turn the JIT compiler off for that, because it does a lot of things that will interfere with that task.

I would consider table and userdata objects with a 'protected' metatable as non-callables and leave a comment about this case.

Reply:
Right now check for `userdata` is updated:

```lua
    local object_metatable = getmetatable(object)
    if (type(object) == 'table' or type(object) == 'userdata')
    and object_metatable and object_metatable.__call then
        -- `__call` metamethod can be only the `function`
        -- and cannot be a `table` | `userdata` | `cdata`
        -- with `__call` methamethod on its own
        return type(object_metatable.__call) == 'function'
    end
```

Reply:
We can detect a protected metatable (at least if `__metatable` is a string) and either accept it or decline. I would accept, because we can't say that it is not a callable (this way is consistent with our approach around cdata). But this should be rare case, so I'll not insist.

Reply:
Fixed:

```diff
-    and object_metatable and object_metatable.__call then
+    if (type(object) == 'table' or type(object) == 'userdata') then
+        -- if metatable type is not `table` -> metatable is protected ->
+        -- cannot detect metamethod `__call` exists
+        if object_metatable and type(object_metatable) ~= 'table' then
+            return true
+        end
+
         -- `__call` metamethod can be only the `function`
         -- and cannot be a `table` | `userdata` | `cdata`
         -- with `__call` methamethod on its own
-        return type(object_metatable.__call) == 'function'
+        if object_metatable and object_metatable.__call then
+            return type(object_metatable.__call) == 'function'
+        end
     end
```

Reply:
Test for this case:

```lua
function g.test_user_sharding_func_protected_metatable()
    -- protected metatable has non-table type -> cannot detect
    -- object is callable-> ok
    local user_sharding_func_name = 'user_sharding_func'
    local user_sharding_func_table = setmetatable({}, {
        __call = function(_, key) return key end, __metatable = 'protected'
    })
    rawset(_G, user_sharding_func_name, user_sharding_func_table)
    g.space.sharding_func = user_sharding_func_name

    local ok, err = ddl.check_schema(g.schema)
    t.assert_equals(err, nil)
    t.assert_equals(ok, true)

    -- protected metatable has table type with __call metamethod -> cannot detect
    -- object is callable -> ok
    local user_sharding_func_name = 'user_sharding_func'
    local user_sharding_func_table = setmetatable({}, {
        __metatable = {__call = function(_, key) return key end}
    })
    rawset(_G, user_sharding_func_name, user_sharding_func_table)
    g.space.sharding_func = user_sharding_func_name

    local ok, err = ddl.check_schema(g.schema)
    t.assert_equals(err, nil)
    t.assert_equals(ok, true)
end
```

Reply:
Ouch, I see now. You're right. It is protected, but using a table in __metatable, so we cannot determine that it is protected.